### PR TITLE
Checkpoint Sync 5/5 - end-to-end test

### PR DESCRIPTION
**Pull request by kasey**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

This PR adds a step to the existing end-to-end test to perform retrieve a checkpoint state+block from one of the running nodes once all nodes are in their final synced state.

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

Due to the way the checkpoint sync period is calculated and the small number of epochs in the end-to-end, this test will currently always fetch the genesis state. It would probably be better for it to fetch HEAD-1, but that would require modifying ws calculation. Alternatively we could tweak the parameters of the checkpoint sync algorithm (like the safety coefficient) to try to shorten the interval to just an epoch or so, but I need to look at the math again to check how many epochs e2e would need to run for this to be feasible.

### CODE REVIEW

**Comment by michaelsproul** [comment]
I just built Prysm using this branch and tried to checkpoint sync against a Lighthouse node, but it failed due to the absence of the `/eth/v1/beacon/weak_subjectivity` endpoint.

```
[2022-03-23 14:39:04]  INFO requesting http://localhost:7052/eth/v1/beacon/weak_subjectivity
[2022-03-23 14:39:05] ERROR main: Error retrieving checkpoint origin state and block: unexpected API response for prysm-only weak subjectivity checkpoint API: code=405, url=http://localhost:7052/eth/v1/beacon/weak_subjectivity, body=response body:
{"code":405,"message":"METHOD_NOT_ALLOWED","stacktraces":[]}: did not receive 2xx response from API
```

I noticed that you try to detect a 404 from the BN and fallback to a backwards-compatible impl if it doesn't exist, but Lighthouse returns a 405 for non-existent HTTP methods (this is our bad, I've opened an issue here to track that: https://github.com/sigp/lighthouse/issues/3112). It might be cool to handle the 405 in the meantime :pray: 

After trying Lighthouse I tried Teku via Infura, and got a `401 Unauthorized`, which I think is due to Prysm not passing along the `project_id:project_secret` (username and password) from the first part of the URL:

```
./beacon-chain --checkpoint-sync-url "https://project_id:project_secret@eth2-beacon-mainnet.infura.io"
...
[2022-03-23 14:47:42]  INFO requesting http://eth2-beacon-mainnet.infura.io/eth/v1/beacon/weak_subjectivity
[2022-03-23 14:47:45] ERROR main: Error retrieving checkpoint origin state and block: unexpected API response for prysm-only weak subjectivity checkpoint API: code=401, url=https://eth2-beacon-mainnet.infura.io:443/eth/v1/beacon/weak_subjectivity, body=response body:
project id required
: did not receive 2xx response from API
```

**Comment by rauljordan** [comment]
Hi @kasey curious about the status of this one and if there's anything I can do to help

**Comment by kasey** [comment]
> I just built Prysm using this branch and tried to checkpoint sync against a Lighthouse node, but it failed due to the absence of the `/eth/v1/beacon/weak_subjectivity` endpoint.
> 
> ```
> [2022-03-23 14:39:04]  INFO requesting http://localhost:7052/eth/v1/beacon/weak_subjectivity
> [2022-03-23 14:39:05] ERROR main: Error retrieving checkpoint origin state and block: unexpected API response for prysm-only weak subjectivity checkpoint API: code=405, url=http://localhost:7052/eth/v1/beacon/weak_subjectivity, body=response body:
> {"code":405,"message":"METHOD_NOT_ALLOWED","stacktraces":[]}: did not receive 2xx response from API
> ```
> 
> I noticed that you try to detect a 404 from the BN and fallback to a backwards-compatible impl if it doesn't exist, but Lighthouse returns a 405 for non-existent HTTP methods (this is our bad, I've opened an issue here to track that: [sigp/lighthouse#3112](https://github.com/sigp/lighthouse/issues/3112)). It might be cool to handle the 405 in the meantime pray
> 
> After trying Lighthouse I tried Teku via Infura, and got a `401 Unauthorized`, which I think is due to Prysm not passing along the `project_id:project_secret` (username and password) from the first part of the URL:
> 
> ```
> ./beacon-chain --checkpoint-sync-url "https://project_id:project_secret@eth2-beacon-mainnet.infura.io"
> ...
> [2022-03-23 14:47:42]  INFO requesting http://eth2-beacon-mainnet.infura.io/eth/v1/beacon/weak_subjectivity
> [2022-03-23 14:47:45] ERROR main: Error retrieving checkpoint origin state and block: unexpected API response for prysm-only weak subjectivity checkpoint API: code=401, url=https://eth2-beacon-mainnet.infura.io:443/eth/v1/beacon/weak_subjectivity, body=response body:
> project id required
> : did not receive 2xx response from API
> ```

Hi Michael, thanks for trying this out! I opened a bug last week on the basic auth issue https://github.com/prysmaticlabs/prysm/issues/10467 and plan to handle 405 equivalent to 404. Sorry this PR got a little stalled, I had to take some personal time off in the last week - I will ping you when these issues are resolved.

File: testing/endtoend/endtoend_test.go [hunk 1]
```diff
@@ -365,19 +404,127 @@ func (r *testRunner) testTxGeneration(ctx context.Context, g *errgroup.Group, ke
...
+	}
+
+	flags := append([]string{}, r.config.BeaconFlags...)
+	flags = append(flags, fmt.Sprintf("--weak-subjectivity-checkpoint=%s", od.CheckpointString()))
+	flags = append(flags, fmt.Sprintf("--checkpoint-state=%s", statePath))
+	flags = append(flags, fmt.Sprintf("--checkpoint-block=%s", blockPath))
+	flags = append(flags, fmt.Sprintf("--genesis-state=%s", genPath))
+
+	//flags = append(flags, fmt.Sprintf("--checkpoint-sync-url=%s", bnAPI))
+	//flags = append(flags, fmt.Sprintf("--genesis-beacon-api-url=%s", bnAPI))
```
Comment (prestonvanloon) [inline testing/endtoend/endtoend_test.go:486]: Do you need these flags?

Comment (kasey) [reply testing/endtoend/endtoend_test.go]: :scissors:

File: beacon-chain/blockchain/process_block.go [hunk 2]
```diff
@@ -498,15 +498,15 @@ func (s *Service) handleEpochBoundary(ctx context.Context, postState state.Beaco
...
-		}
 		copied := postState.Copy()
 		copied, err := transition.ProcessSlots(ctx, copied, copied.Slot()+1)
 		if err != nil {
 			return err
 		}
+		// Update caches for the next epoch at epoch boundary slot - 1.
+		if err := helpers.UpdateCommitteeCache(copied, coreTime.CurrentEpoch(copied)); err != nil {
+			return err
+		}
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:509]: It'll be cleaner to move all the cache-related changes to another PR. I think that will be easier and safer for a detailed review

File: beacon-chain/blockchain/execution_engine.go [hunk 3]
```diff
@@ -102,6 +102,58 @@ func (s *Service) notifyForkchoiceUpdate(ctx context.Context, headState state.Be
...
+	b, err = s.cfg.BeaconDB.Block(ctx, r)
+	if err != nil {
+		return [32]byte{}, errors.Wrap(err, "could not get finalized block")
+	}
+	if b != nil {
+		return getPayloadHash(b.Block())
+	}
+
+	return [32]byte{}, errors.Errorf("finalized block with root %#x does not exist in the db or our cache", s.ensureRootNotZeros(finalizedRoot))
+}
```
[the remark this answers is not in the document]
Comment (kasey) [reply beacon-chain/blockchain/execution_engine.go]: Is it possible these fixes resolved the problem? https://github.com/prysmaticlabs/prysm/pull/10630/files#diff-c4cc28786298264208b7c0d5ac74efa77f38e3c754f48066bf22145374542563

What's the smallest test you can think of implementing to demonstrate this problem without the changes in this branch?

File: beacon-chain/blockchain/process_block.go (hunk 2 above)
Comment (kasey) [reply beacon-chain/blockchain/process_block.go:498]: see https://github.com/prysmaticlabs/prysm/pull/10639

**Comment by kasey** [comment]
@michaelsproul we just merged PR #10723 which significantly changes the way our checkpoint sync works. Instead of trying to fetch the block from the beginning of the latest weak subjectivity period, we simply fetch the latest finalized block, working on the assumption that this is preferable in terms of sync speed. This is also more compatible with Infura, which fails to serve the state-by-slot request we need to accompany the checkpoint block, but does a fine job fetching `genesis` and `finalized` states + block by root. I was convinced by Adrian Sutton that this is generally just as safe as the other method given that users can confirm the fetched roots are canonical using a trusted beacon API. Please give this a try and let me know how it works for you.

**Comment by kasey** [comment]
Closing this PR as checkpoint sync behavior changed heavily in https://github.com/prysmaticlabs/prysm/pull/10723 which includes its own e2e test.

**Pull request closed without merging**